sql-plugin/src/main/scala/com/nvidia/spark/rapids/CostBasedOptimizer.scala

```diff
@@ -270,6 +273,62 @@ class DefaultCostModel(conf: RapidsConf) extends CostModel {
 
 }
 
+/**
+ * Estimate the number of rows that an operator will output. Note that these row counts are
+ * the aggregate across all partitions.
+ *
+ * Logic is based on Spark's SizeInBytesOnlyStatsPlanVisitor. which operates on logical plans
+ * and only computes data sizes, not row counts.
+ */
+object RowCountPlanVisitor {
+
+  def visit(plan: SparkPlanMeta[_]): Option[BigInt] = plan.wrapped match {
+    case p: QueryStageExec =>
+      ShimLoader.getSparkShims.getQueryStageRuntimeStatistics(p).rowCount
+    case GlobalLimitExec(limit, _) =>
+      visit(plan.childPlans.head).map(_.min(limit)).orElse(Some(limit))
+    case LocalLimitExec(limit, _) =>
+      visit(plan.childPlans.head).map(_.min(limit)).orElse(Some(limit))
+    case p: SortMergeJoinExec =>
```

Comment:
I would add in a TakeOrderedAndProject in here that behaves a lot like the limits, but it also sorts the data.

Reply:
Added